From f884f7855d610a8f61a277206c9c443e481f274c Mon Sep 17 00:00:00 2001
From: Alejandro Abdelnur <tucu@apache.org>
Date: Mon, 27 Aug 2012 22:38:13 +0000
Subject: [PATCH 1180/1357] MR1: MAPREDUCE-4595. TestLostTracker failing - possibly due to a race in JobHistory.JobHistoryFilesManager#run() (kkambatl via tucu)

git-svn-id: https://svn.apache.org/repos/asf/hadoop/common/branches/branch-1@1377895 13f79535-47bb-0310-9956-ffa450edef68
(cherry picked from commit 327f6378f2cdb9923b2b54a3ef39f8b207ff8173)

Reason: TestLostTracker fails due to a race
Ref: CDH-7104
Author: Karthik Kambatla
---
 .../org/apache/hadoop/mapred/TestLostTracker.java  |   19 +++++++++++++++++++
 1 files changed, 19 insertions(+), 0 deletions(-)

diff --git a/src/test/org/apache/hadoop/mapred/TestLostTracker.java b/src/test/org/apache/hadoop/mapred/TestLostTracker.java
index e69b157..2b46506 100644
--- a/src/test/org/apache/hadoop/mapred/TestLostTracker.java
+++ b/src/test/org/apache/hadoop/mapred/TestLostTracker.java
@@ -102,6 +102,25 @@ public class TestLostTracker extends TestCase {
       testTaskStatuses(mtip.getTaskStatuses());
     }
     
+    // Before validating job history, wait for the history file to be available
+    JobInProgress jip = mr.getJobTrackerRunner().getJobTracker().getJob(id);
+    long beginWaiting = System.currentTimeMillis();
+    final long MAX_WAIT_TIME = 5 * 60 * 1000;
+    while (System.currentTimeMillis() - beginWaiting < MAX_WAIT_TIME) {
+      if (!jip.getHistoryFile().equals("")) {
+        break;
+      } else {
+        try {
+          Thread.sleep(50);
+        } catch (InterruptedException ie) {
+          // do nothing
+        }
+      }
+    }
+
+    assertFalse("Job history file needs to be set for further validation", jip
+        .getHistoryFile().equals(""));
+
     // validate the history file
     TestJobHistory.validateJobHistoryFileFormat(id, job, "SUCCESS", true);
     TestJobHistory.validateJobHistoryFileContent(mr, rJob, job);
-- 
1.7.0.4

